### PR TITLE
Add support for SyncConfiguration.assetFile

### PR DESCRIPTION
**Pull request by cmelchior**
Closes https://github.com/realm/realm-java/issues/7494

WIP, currently blocked on https://github.com/realm/realm-core/pull/4878

### CODE REVIEW

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt [hunk 1]
```diff
@@ -804,6 +806,174 @@ class SyncedRealmTests {
...
+
+    @Test
+    fun assetFile() {
+        val user = createNewUser()
+        val config = SyncConfiguration.Builder(user, "e873fb25-11ef-498f-9782-3c8e1cd2a12c")
+            .assetFile("synced_realm_e873fb25-11ef-498f-9782-3c8e1cd2a12c_no_client_id.realm")
+            .schema(SyncColor::class.java)
+            .build()
+
+        Realm.getInstance(config).use { realm: Realm ->
```
Comment (cmelchior) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt:940]: I think this also needs to verify you can call `SyncSession.uploadAllLocalChanges()` ... at least that was one of the bugs in Core, that you could open the file locally, but you could not sync with it.

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 2]
```diff
@@ -300,7 +300,9 @@ protected void removeAllListeners() {
     }
 
     /**
-     * Writes a compacted copy of the Realm to the given destination File.
+     * Writes a compacted copy of the Realm to the given destination File. On a synced Realm it will
+     * write a copy stripping the client file identifier that will allow to reuse the Realm without
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java:304]: `without redownloading the realm` sounds a bit funny...maybe rephrase to `resue the Realm on another device`?

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java [hunk 3]
```diff
@@ -199,20 +199,22 @@ protected RealmProxyMediator getSchemaMediator() {
     }
 
     /**
-     * Indicates if there is available asset file for copy action.
+     * Indicates if an asset file has configured for this configuration.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java:202]:
```suggestion
     * Indicates if an asset file has been configured for this configuration.
```

**Review by cmelchior — COMMENTED** [review]
Nice, thanks for getting it over the finish line 👏 

Only two things missing:
- Entry in the changelog
- A small update to one of the tests

But otherwise feel free to merge.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt (hunk 1 above)
Comment (clementetb) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncedRealmTests.kt:940]: I removed `uploadAllLocalChanges()` because I though it was not supported to run it at that point, but it's a deeper issue around how we test `assetFile()`.

The sync Realm `writeCopy` file is no longer bound to the user, but it is still to the server instance: the asset file we provide only works on the instance it was created and will not work on other instances, thats why `uploadAllLocalChanges` fails.

We might need to mock the assets to be able to pass a freshly generated one into the config.

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java (hunk 2 above)
Comment (clementetb) [reply realm/realm-library/src/main/java/io/realm/BaseRealm.java:304]: Maybe:
`Writes a compacted copy of the Realm to the given destination File. The resulting file can be used as initial dataset to bootstrap local or synced Realm in other devices.`

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 ### Enhancements
 * [RealmApp] Reduced native memory usage when working with synchronized Realms.
+* [RealmApp] Added support for `writeCopyTo` and `RealmConfiguration.Builder.assetFile()` on synchronized Realms.
```
Comment (cmelchior) [inline CHANGELOG.md:5]:
```suggestion
* [RealmApp] Added support for `Realm.writeCopyTo()` and `SyncConfiguration.Builder.assetFile()` on synchronized Realms.
```
